src/v2i-hub/TelematicBridgePlugin/src/TelematicUnit.cpp

```diff
@@ -0,0 +1,314 @@
+                {
+                    obj->addSelectedTopic(itr->asString());
+                }
+            }
+            string reply = "request received!";
+            auto s = natsConnection_PublishString(nc, natsMsg_GetReply(msg), reply.c_str());
+            natsMsg_Destroy(msg);
+            if (s == NATS_OK)
+            {
+                PLOG(logDEBUG3) << "Selected topics replied: " << reply;
+            }
+        }
+    }
+
+    void TelematicUnit::onCheckStatusCallback(natsConnection *nc, natsSubscription *sub, natsMsg *msg, void *object)
+    {
+        if (natsMsg_GetReply(msg) != nullptr)
+        {
+            auto s = natsConnection_PublishString(nc, natsMsg_GetReply(msg), "OK");
+            if (s == NATS_OK)
+            {
+                PLOG(logDEBUG3) << "Received check status msg: " << natsMsg_GetSubject(msg) << " " << natsMsg_GetData(msg) << ". Replied: OK";
+            }
+            natsMsg_Destroy(msg);
+        }
+    }
+
+    string TelematicUnit::constructPublishedDataString(const unit_st &unit, const string &eventLocation, const string &testingType, const string &eventName, const string &topicName, const Json::Value &payload) const
+    {
+        Json::Value message;
+        message[UNIT_ID_KEY] = unit.unitId;
+        message[UNIT_NAME_KEY] = unit.unitName;
+        message[UNIT_TYPE_KEY] = unit.unitType;
+        message[LOCATION_KEY] = eventLocation;
+        message[TESTING_TYPE_KEY] = testingType;
+        message[EVENT_NAME_KEY] = eventName;
+        message[TOPIC_NAME_KEY] = topicName;
+        message[TIMESTAMP_KEY] = payload.isMember("timestamp") ? payload["timestamp"].asUInt64() * MILLI_TO_MICRO : duration_cast<microseconds>(system_clock::now().time_since_epoch()).count();
+        message[PAYLOAD_KEY] = payload;
+        Json::FastWriter fasterWirter;
+        string jsonStr = fasterWirter.write(message);
+        return jsonStr;
+    }
+
+    Json::Value TelematicUnit::parseJson(const string &jsonStr)
+    {
+        Json::Value root;
+        Json::Reader reader;
+        bool parsingSuccessful = reader.parse(jsonStr, root);
+        if (!parsingSuccessful)
+        {
+            throw TelematicBridgeException("Error parsing the reply message");
+        }
+        return root;
+    }
+
+    string TelematicUnit::constructAvailableTopicsReplyString(const unit_st &unit, const string &eventLocation, const string &testingType, const string &eventName, const vector<string> &availableTopicList, const string &excludedTopics)
+    {
+        Json::Value message;
+        message[UNIT_ID_KEY] = unit.unitId;
+        message[UNIT_NAME_KEY] = unit.unitName;
+        message[UNIT_TYPE_KEY] = unit.unitType;
+        message[LOCATION_KEY] = eventLocation;
+        message[TESTING_TYPE_KEY] = testingType;
+        message[EVENT_NAME_KEY] = eventName;
+        message[TIMESTAMP_KEY] = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
```

Comment:
If you do not extend PluginClientClockAware and use the carma-clock object this plugin will not support simulation.

Reply:
That is different case. We will address this later when we integrate this with simulation.